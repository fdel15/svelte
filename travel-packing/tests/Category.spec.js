import { cleanup, fireEvent, render, waitFor } from '@testing-library/svelte';
import Category from '../src/Category.svelte';

describe('Category', () => {
    let itemCount = 0; // Will be set by beforeEach function

    const dragAndDrop = {}
    const category = { id: 1, name: 'Stuff', items: {} };
    const categories = [category]
    const show = 'all';
    const props = { categories, category, show, dragAndDrop };

    beforeEach(() => {
        category.items = {
            1: { id: 1, name: 'socks', packed: true },
            2: { id: 2, name: 'shoes', packed: false }
        };
        itemCount = Object.keys(category.items).length;
    })

    afterEach(cleanup);

    test('should match snapshot', async() => {
        const container = render(Category, props);
        expect(container).toMatchSnapshot();
    })

    function expectItemCount(count) {
        return waitFor(() => {
            const lis = document.querySelectorAll('li');
            expect(lis.length).toBe(count);
        })
    };

    test('should render', async() => {
        const { getByText } = render(Category, props);
        expect(getByText('Stuff'));
        expect(getByText('1 of 2 remaining'));
        expect(getByText('New Item'));
        expect(getByText('Add Item'));
        await expectItemCount(itemCount);
    });

    test('should add an item', async() => {
        const { getByText, getByTestId } = render(Category, props);
        const input = getByTestId('item-input');
        const value = 't-shirts';

        fireEvent.input(input, { target: { value } }); // populates input
        fireEvent.click(getByText('Add Item'));

        await expectItemCount(itemCount + 1); // element was added
        expect(getByText(value)); // element appears in the dom
    });

    test('should delete an item', async() => {
        const { getAllByTestId } = render(Category, props);
        const delteBtns = getAllByTestId('delete') // each item will have a delete button

        fireEvent.click(delteBtns[0]) // click first delete button
        await expectItemCount(itemCount - 1); // element was added
    })

    test('should toggle an item', async() => {
        const { container, getByText } = render(Category, props);

        const checkboxes = document.querySelectorAll('input[type="checkbox"]');
        expect(checkboxes.length).toBe(2);

        const [shoesCheckbox, socksCheckBox] = checkboxes; // specific order is set by sorting by name

        expect(socksCheckBox.nextElementSibling.textContent).toBe('socks'); // verifies the order from hardcoded
        await fireEvent.click(socksCheckBox);
        expect(getByText('2 of 2 remaining')) // now nothing in the category is packed

        expect(shoesCheckbox.nextElementSibling.textContent).toBe('shoes');
        await fireEvent.click(shoesCheckbox);
        expect(getByText('1 of 2 remaining'));

    })
})