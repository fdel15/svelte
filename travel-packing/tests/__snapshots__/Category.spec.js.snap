// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category should match snapshot 1`] = `
Object {
  "component": Category {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Object {
          "id": 1,
          "items": Object {
            "1": Object {
              "id": 1,
              "name": "socks",
              "packed": true,
            },
            "2": Object {
              "id": 2,
              "name": "shoes",
              "packed": false,
            },
          },
          "name": "Stuff",
        },
        Object {},
        <dialog
          class="dialog svelte-z0mxa4"
        >
          <header
            class="svelte-z0mxa4"
          >
             
            <div
              class="title svelte-z0mxa4"
            >
              Category
            </div>
             
            <button
              class="close-btn svelte-z0mxa4"
            >
              âœ–
            </button>
          </header>
           
          <main
            class="svelte-z0mxa4"
          >
            <div>
              
            </div>
          </main>
        </dialog>,
        false,
        "",
        "",
        false,
        "1 of 2 remaining",
        Array [
          Object {
            "id": 2,
            "name": "shoes",
            "packed": false,
          },
          Object {
            "id": 1,
            "name": "socks",
            "packed": true,
          },
        ],
        [Function],
        Object {
          "duration": 700,
          "easing": [Function],
          "times": 2,
        },
        [Function],
        [Function],
        Array [
          Object {
            "id": 1,
            "items": Object {
              "1": Object {
                "id": 1,
                "name": "socks",
                "packed": true,
              },
              "2": Object {
                "id": 2,
                "name": "shoes",
                "packed": false,
              },
            },
            "name": "Stuff",
          },
        ],
        "all",
        Array [
          Object {
            "id": 1,
            "name": "socks",
            "packed": true,
          },
          Object {
            "id": 2,
            "name": "shoes",
            "packed": false,
          },
        ],
        1,
        2,
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
      ],
      "on_mount": Array [],
      "props": Object {
        "categories": 13,
        "category": 0,
        "dragAndDrop": 1,
        "show": 14,
      },
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <section
        class="svelte-1rk183a"
      >
        <h3
          class="svelte-1rk183a"
        >
          <span>
            Stuff
          </span>
           
          <span
            class="status svelte-1rk183a"
          >
            1 of 2 remaining
          </span>
           
          <button
            class="icon svelte-1rk183a"
          >
            ðŸ—‘
          </button>
        </h3>
         
        <form>
          <label>
            New Item
      
            <input
              class="svelte-1rk183a"
              data-testid="item-input"
              required=""
            />
          </label>
           
          <button
            class="svelte-1rk183a"
          >
            Add Item
          </button>
        </form>
         
        <ul
          class="svelte-1rk183a"
        >
          <div>
            <li
              class="svelte-1fnd29c"
              draggable="true"
            >
              <input
                class="svelte-1fnd29c"
                type="checkbox"
              />
               
              <span
                class="packed-false svelte-1fnd29c"
              >
                shoes
              </span>
               
              <button
                class="icon svelte-1fnd29c"
                data-testid="delete"
              >
                ðŸ—‘
              </button>
            </li>
             
          </div>
          <div>
            <li
              class="svelte-1fnd29c"
              draggable="true"
            >
              <input
                class="svelte-1fnd29c"
                type="checkbox"
              />
               
              <span
                class="packed-true svelte-1fnd29c"
              >
                socks
              </span>
               
              <button
                class="icon svelte-1fnd29c"
                data-testid="delete"
              >
                ðŸ—‘
              </button>
            </li>
             
          </div>
        </ul>
      </section>
       
      <dialog
        class="dialog svelte-z0mxa4"
      >
        <header
          class="svelte-z0mxa4"
        >
           
          <div
            class="title svelte-z0mxa4"
          >
            Category
          </div>
           
          <button
            class="close-btn svelte-z0mxa4"
          >
            âœ–
          </button>
        </header>
         
        <main
          class="svelte-z0mxa4"
        >
          <div>
            
          </div>
        </main>
      </dialog>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
